src/transaction/log_manager.c

```diff
@@ -14684,6 +14733,457 @@ cdc_is_filtered_user (char *user)
   return false;
 }
 
+static int
+flashback_find_start_lsa (THREAD_ENTRY * thread_p, FLASHBACK_LOGINFO_CONTEXT * context)
+{
+  LOG_PAGE *log_page_p = NULL;
+  char log_pgbuf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+
+  int error = NO_ERROR;
+
+  LOG_LSA process_lsa;
+
+  LOG_RECORD_HEADER *log_rec_header;
+
+  log_page_p = (LOG_PAGE *) PTR_ALIGN (log_pgbuf, MAX_ALIGNMENT);
+
+  assert (!LSA_ISNULL (&context->end_lsa));
+  LSA_COPY (&process_lsa, &context->end_lsa);
+
+  if (logpb_fetch_page (thread_p, &process_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+    {
+      /* can not find the log page */
+      error = ER_FAILED;
+      goto exit;
+    }
+
+begin:
+  log_rec_header = LOG_GET_LOG_RECORD_HEADER (log_page_p, &process_lsa);
+
+  if (LSA_ISNULL (&log_rec_header->prev_tranlsa))
+    {
+      LSA_COPY (&context->start_lsa, &process_lsa);
+      error = NO_ERROR;
+
+      goto exit;
+    }
+  else
+    {
+      LSA_COPY (&process_lsa, &log_rec_header->prev_tranlsa);
+    }
+
+  if (log_page_p->hdr.logical_pageid != process_lsa.pageid)
+    {
+      if (logpb_is_page_in_archive (process_lsa.pageid))
+	{
+	  int arv_num = 0;
+
+	  if (logpb_fetch_from_archive (thread_p, process_lsa.pageid, log_page_p, &arv_num, NULL, false) == NULL)
+	    {
+	      /* can not find the archive log volume */
+	      error = ER_FAILED;
+	      goto exit;
+	    }
+	}
+      else
+	{
+	  if (logpb_fetch_page (thread_p, &process_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+	    {
+	      error = ER_FAILED;
+	      goto exit;
+	    }
+	}
+    }
+
+  goto begin;
+
+exit:
+  return error;
+}
+
+static bool
+flashback_class_filter (FLASHBACK_LOGINFO_CONTEXT * context, OID classoid)
+{
+  for (int i = 0; i < context->num_class; i++)
+    {
+      if (OID_EQ (&context->classlist[i], &classoid))
+	{
+	  return true;
+	}
+    }
+
+  return false;
+}
+
+int
+flashback_make_loginfo (THREAD_ENTRY * thread_p, FLASHBACK_LOGINFO_CONTEXT * context)
+{
+  LOG_LSA cur_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA next_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA process_lsa = LSA_INITIALIZER;
+
+  LOG_PAGE *log_page_p = NULL;
+  char log_pgbuf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+
+  LOG_RECORD_HEADER *log_rec_header;
+
+  char tran_user[DB_MAX_USER_LENGTH + 1] = "\0";
+  int trid;
+
+  int error = NO_ERROR;
+
+  LOG_RECTYPE log_type;
+
+  char *supplement_data = NULL;
+
+  RECDES supp_recdes = RECDES_INITIALIZER;
+  RECDES undo_recdes = RECDES_INITIALIZER;
+  RECDES redo_recdes = RECDES_INITIALIZER;
+
+  int num_loginfo = 0;
+  CDC_LOGINFO_ENTRY *log_info_entry = NULL;
+
+  if (LSA_ISNULL (&context->start_lsa))
+    {
+      if ((error = flashback_find_start_lsa (thread_p, context)) != NO_ERROR)
```

Comment:
yes, that can be done only for when direction is forward, and just for the first time user request. 

this is processed only for rare case that transaction is started out of the time range (transaction start time > flashback user input start time).

Reply:
Yes, I agree with you.
I will leave this one to refactor. 
it may be modified to `flashback_find_start_lsa (thread_p, log_page_p, context)` to use `log_page_p` which is fetched in the function.

Reply:
OK.